### DISCUSSION TITLE
Question about performance dev/prod

### DISCUSSION THREAD

**Discussion by cadilhac** [ravendb/ravendb#17638] [Q&A]
Hi,

I have a dev machine (recent and reasonably powerful - Ryzen 5 5600X with 32 Gb and good SSDs) with 2 different possible setups. RavenDb can be run as a simple windows exe or inside a linux VM. VS runs on the same machine.

The production environment consists of 2 servers on the free offering of Oracle cloud. One server hosts the web server. The other one hosts the RavenDb instance.

My web app shows a map with a bunch of markers reprensenting users in the database. For this test, the app queries the db which returns around 150 records. When I check the ajax get queries in Chrome, they last:

1. Around 120 ms in production
2. Around 1.6 s in development (web server ran in debug or release mode, RavenDb on windows or in the linux vm)

I'm quite surprised by this result. I would have thought that it would go blazingly fast on my dev machine... Any idea why it's not the case?

**Comment by garayx** [comment]
Hi

The question is how do you measure the latency ?
Do you create a new document store each time and then send a query or you reuse the same document store ?
Do you use caching ?

In general you can open the traffic watch view in studio, fire the request, and you will see the actual duration (in ms) of the request on the RavenDB server:

![image](https://github.com/ravendb/ravendb/assets/16625387/b2a9da57-fa25-43d8-ab0d-1e4e373be78c)

**Reply by cadilhac** [reply]
Thank you. You are right, in traffic watch the duration of the request is very short. So my problem is not on the RavenDb server. The problem is then on my web server with the RavenDb client. Here is why:

My initial spatial query is very short as just said. In this query I have an Include<OtherClass>(x => x.otherId). I collect all these otherIds in an array (100 items) and I then call:

````
await RavenSession.LoadAsync<OtherClass>(otherIds);
````

This is this call that takes almost 2 seconds. I checked in traffic watch that it doesn't generate a call to the RavenDb server (since Include was used). So everything is in the RavenDb session. So why is this so long?
I tried 2 other things:

1. Instead of a LoadAsync call, I used a Query with an In clause. Same duration
2. Instead of a single LoadAsync class, I used as many as otherIds. Same duration
3. I removed the Include class. Therefore the LoasAsync goes to the server. About the same duration o even a bit shorter

So, although the entities are in the session cache thanks to the Include call, it's as if it had no effect. My LoadAsync call does not go to the server (checked in traffic watch) but it takes as long...

PS: document store created only once in the web app. Each web request has a single opened session. Also note that each entity has a very small size.

**Reply by ayende** [reply]
Check if you are doing something in the constructor or deserialization of those entities.
Try running this in the profiler

**Reply by cadilhac** [reply]
No, this is a very basic entity class.

> Try running this in the profiler

Please, explain. This is a single call and I measured its duration. What more do you mean?

**Reply by ayende** [reply]
Under the profiler, you can see _why_ it is taking so long. If this is an in memory only operation, that may explain the cost.
Usually you can see a single big operation.

For example, if the size of the document on the dev env is 100MB, it may be very fast on the network, but slow to deserialize.

**Reply by cadilhac** [reply]
I will check this profiler stuff tomorrow but what I can already say is that sizes are very small here. In prod or env, they are the same. The spatial query (with the include) returns 156K of data, that's all.

**Reply by cadilhac** [reply]
I'm puzzled by the profiler test. When I run the profiler (with instrumentation of my code), still running in debug, the problematic call (LoadAsync) becomes fast. I don't even need to know how long it takes. I mean that from the user point of view, the web page shows the result as I'm used to see it in prod, i.e. fast. Back to running without the profiler, the call to LoadAsync is slow again...

**Reply by ppekrol** [reply]
Is there a chance that in your setup there is some DNS problem for example? Antivirus? Proxy?

**Reply by ayende** [reply]
That is a strong indication that this may not be it that is causing it.
Are you running this in debug _inside_ the debugger?

**Reply by cadilhac** [reply]
> Is there a chance that in your setup there is some DNS problem for example? Antivirus? Proxy?

Not that I know. I turned off the antivirus (malwarebytes) to see. No changes.

> Are you running this in debug inside the debugger?

Yes. Always debug mode.

**Reply by ayende** [reply]
What happens if you are running in debug mode but outside the profiler?

**Reply by cadilhac** [reply]
This is what I did from the beginning. Always debug mode. So, if outside profiler, call is long.
